### PR TITLE
Unbeatable player

### PR DESCRIPTION
**Pull request by andrewMacmurray**
+ adds unbeatable player
+ validates input from user
+ can choose different game options

### CODE REVIEW

File: lib/console/console.rb [hunk 1]
```diff
@@ -13,44 +12,94 @@ def greet_user
...
+    game_instructions(players.current_player_symbol)
+    players
+  end
+
+  def request_move(board = nil)
+    prompt
+    get_move
+  end
+
+  def move_summary(move, board, player, oponent)
```
Comment (RabeaGleissner) [inline lib/console/console.rb]: I think "opponent" is spelled with double "p"!

File: lib/console/console.rb [hunk 2]
```diff
@@ -13,44 +12,94 @@ def greet_user
...
+  def print_board(board)
     board_string = board_renderer.render(board)
     print(board_string)
   end
 
-  def show_move_summary(move, player, oponent)
-    print(messages.player_move(move, player))
-    print(messages.player_turn(oponent))
+  def print_terminus(board)
+    print_draw    if board.full?
```
Comment (RabeaGleissner) [inline lib/console/console.rb]: This reads very nicely!
But couldn't it be that the board is full but there is still a winner?

File: lib/console/console_io.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  def puts(message)
+    output.puts(message)
+  end
+
+  def print(message)
+    output.print(message)
+  end
+
+  def read_int_in_range(from, to)
+    val = get_next
```
Comment (RabeaGleissner) [inline lib/console/console_io.rb]: I think it's better not to use shortened variable names... do you mean "value" in this case? Or "valid"?

Comment (RabeaGleissner) [reply lib/console/console_io.rb]: Maybe "move" would be a better name?

File: lib/console/messages.rb [hunk 4]
```diff
@@ -11,6 +11,10 @@ def options
     ]
   end
 
+  def prompt() "> " end
+
+  def unrecognised() "Sorry I didn't recognise that" end
```
Comment (RabeaGleissner) [inline lib/console/messages.rb]: I find it a bit unusual to have a method on one line like this! Probably better to have it over three lines, even though it's short.

File: lib/core/game.rb [hunk 5]
```diff
@@ -1,49 +1,49 @@
...
+    while !board.terminus_reached? do
+      eval_move!
+    end    
+    ui.game_summary(board)
   end
 
-  def make_move(move)
-    board.make_move(move, players.current_player_symbol)
-    ui.show_board(board)
+  def eval_move!
```
Comment (RabeaGleissner) [inline lib/core/game.rb]: same here... I think "evaluate_move" is better than the shortened method name

File: spec/console/console_io_spec.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  it "reads an integer in a given range" do
+    input = StringIO.new("3")
+    output = StringIO.new
+    console_io = ConsoleIO.new(input: input, output: output)
+
+    result = console_io.read_int_in_range(1, 9)
+    expect(result).to eq(3)
+  end
+
+  it """prints an error message for non integer input
```
Comment (RabeaGleissner) [inline spec/console/console_io_spec.rb]: any reason for the double double quotes? 😂

File: spec/core/computer_player_spec.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  end
+
+  it "takes top left corner if not already taken" do
+    player = ComputerPlayer.new(:X)
+    board = Board.new
+
+    expect(player.request_move(board)).to eq(1)
+  end
+
+  it "request move via minimax" do
```
Comment (RabeaGleissner) [inline spec/core/computer_player_spec.rb]: If this method tests that the move is requested via minimax, how are you asserting that? When I read the test name I expected that you would use a spy to assert that.
Maybe the test name should just be "computer makes a move" or something like that?

File: spec/core/human_player_spec.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+require "console/console"
+require "console/console_io"
+
+describe HumanPlayer do
+  it "has correct symbol" do
+    player = HumanPlayer.new(:X, build_ui)
+    expect(player.symbol).to eq(:X)
+  end
+
+  it "requests a move via the ui" do
```
Comment (RabeaGleissner) [inline spec/core/human_player_spec.rb:12]: same here, the test name sounds to me like you would be spying on the Ui

**Review by RabeaGleissner — COMMENTED** [review]
Nice job! 👍

File: lib/console/console.rb (hunk 1 above)
Comment (andrewMacmurray) [reply lib/console/console.rb]: whoops 🙈

File: lib/console/console_io.rb (hunk 3 above)
Comment (andrewMacmurray) [reply lib/console/console_io.rb]: Good point, will change it to value

File: lib/console/console.rb (hunk 2 above)
Comment (andrewMacmurray) [reply lib/console/console.rb]: Thanks! But yes that line should be last, will add a test case for it

File: spec/console/console_io_spec.rb (hunk 6 above)
Comment (andrewMacmurray) [reply spec/console/console_io_spec.rb]: lol it was for a multiline string, but maybe I don't need them (also probably a sign the test / method is doing too much)

**Comment by andrewMacmurray** [comment]
Thanks @RabeaGleissner!

**Pull request merged**